### PR TITLE
all: add signout redirect url

### LINKED ISSUE #1213: Make dashboard Sign Out destination configurable
**Is your feature request related to a problem? Please describe.**

It would be helpful if the destination for redirection after sign out via the dashboard (`https://proxy-service/.pomerium`) were configurable.

**Describe the solution you'd like**

Configuration option for the proxy service's dashboard to make the redirection after sign out configurable.

**Describe alternatives you've considered**

Not sure what alternatives might exist here?

**Explain any additional use-cases**

None yet.

**Additional context**

N/a

### PR DESCRIPTION
**Pull request by cuonglm**
## Summary
Add signout redirect url config, allow user to config where they will be redirected to after signing out.

## Related issues
Fixes #1213

**Checklist**:
- [x] add related issues
- [x] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

File: config/options.go [hunk 1]
```diff
@@ -112,6 +112,10 @@ type Options struct {
 	// used for authentication requests and callbacks
 	AuthenticateURLString string   `mapstructure:"authenticate_service_url" yaml:"authenticate_service_url,omitempty"`
 	AuthenticateURL       *url.URL `yaml:"-,omitempty"`
+	// AuthenticateURL represents the externally accessible http endpoints
```
Comment (desimone) [inline config/options.go:115]: I think you forgot to update the comment here?

File: config/options.go [hunk 2]
```diff
@@ -539,6 +543,14 @@ func (o *Options) Validate() error {
 		o.AuthenticateURL = u
 	}
 
+	if o.SignOutRedirectURLString != "" {
```
Comment (desimone) [inline config/options.go:546]: Can we do this validation check in authenticate only?

File: go.sum [hunk 3]
```diff
@@ -270,15 +270,11 @@ github.com/gophercloud/gophercloud v0.3.0/go.mod h1:vxM41WHh5uqHVBMZHzuwNOHh8XEo
 github.com/gopherjs/gopherjs v0.0.0-20181017120253-0766667cb4d1 h1:EGx4pi6eqNxGaHF6qqu48+N2wcFQ5qg5FXgOdqsJ5d8=
 github.com/gopherjs/gopherjs v0.0.0-20181017120253-0766667cb4d1/go.mod h1:wJfORRmW1u3UXTncJ5qlYoELFm8eSnnEO6hX4iZ3EWY=
 github.com/gorilla/context v1.1.1/go.mod h1:kBGZzfjB9CEq2AlWe17Uuf7NDRt0dE0s8S51q0aT7Yg=
-github.com/gorilla/handlers v1.4.2 h1:0QniY0USkHQ1RGCLfKxeNHK9bkDHGRYGNDFBCS+YARg=
```
Comment (desimone) [inline go.sum:273]: can you merge with latest and re-run mod tidy to fix any weirdness from the set of renovate bot merges?

File: docs/reference/readme.md [hunk 4]
```diff
@@ -601,6 +601,16 @@ See also:
...
 
+### Signout Redirect URL
+
+- Environmental Variable: `SIGNOUT_REDIRECT_URL`
+- Config File Key: `signout_redirect_url`
+- Type: `URL`
+- Required
+- Example: `https://signout-redirect-url.corp.example.com`
+
+Signout redirect url is the url user will be redirected to after signing out.
```
Comment (desimone) [inline docs/reference/readme.md:612]: Can we also add a note here saying it is also possible set a redirect url by passing in a custom queryparam / post value when hitting the `./pomerium/signout` endpoint?

File: docs/reference/readme.md [hunk 5]
```diff
@@ -601,6 +601,16 @@ See also:
 
 Authenticate Service URL is the externally accessible URL for the authenticate service.
 
+### Signout Redirect URL
```
Comment (desimone) [inline docs/reference/readme.md:604]: Can you put in alphabetic order?

File: authenticate/handlers.go [hunk 6]
```diff
@@ -282,6 +282,9 @@ func (a *Authenticate) SignOut(w http.ResponseWriter, r *http.Request) error {
 	// no matter what happens, we want to clear the session store
 	state.sessionStore.ClearSession(w, r)
 	redirectString := r.FormValue(urlutil.QueryRedirectURI)
+	if sru := a.options.Load().SignOutRedirectURL; sru != nil {
+		redirectString = sru.String()
```
Comment (desimone) [inline authenticate/handlers.go:286]: If both are supplied, I think query param should win?

File: config/options.go (hunk 1 above)
Comment (cuonglm) [reply config/options.go:115]: Ops, my bad. Fixed.

File: config/options.go (hunk 2 above)
Comment (cuonglm) [reply config/options.go:546]: I think no, because we should also use it in proxy, let see my latest commit.

File: go.sum (hunk 3 above)
Comment (cuonglm) [reply go.sum:273]: I did, renovate bot seems never run `go mod tidy` :(

File: docs/reference/readme.md (hunk 4 above)
Comment (cuonglm) [reply docs/reference/readme.md:612]: ok

File: docs/reference/readme.md (hunk 5 above)
Comment (cuonglm) [reply docs/reference/readme.md:604]: Fixed.

File: authenticate/handlers.go (hunk 6 above)
Comment (cuonglm) [reply authenticate/handlers.go:286]: Ah yes.

File: go.sum (hunk 3 above)
Comment (desimone) [reply go.sum:273]: https://github.com/pomerium/pomerium/blob/3d7206dc1e61864c9ed94c45abdf1c6f20440b06/renovate.json#L12

So weird.

File: config/options.go [hunk 7]
```diff
@@ -112,6 +112,9 @@ type Options struct {
 	// used for authentication requests and callbacks
 	AuthenticateURLString string   `mapstructure:"authenticate_service_url" yaml:"authenticate_service_url,omitempty"`
 	AuthenticateURL       *url.URL `yaml:"-,omitempty"`
+	// SignOutRedirectURL represents the url that user will be redirected to after signing out.
```
Comment (desimone) [inline config/options.go:115]:
```suggestion
	// SignOutRedirectURL represents the url that a user will be redirected to after signing out.
```

File: docs/reference/readme.md [hunk 8]
```diff
@@ -1010,6 +1010,19 @@ Use this token to authenticate requests to a Kubernetes API server.
...
+
+- Environmental Variable: `SIGNOUT_REDIRECT_URL`
+- Config File Key: `signout_redirect_url`
+- Type: `URL`
+- Required
+- Example: `https://signout-redirect-url.corp.example.com`
+
+Signout redirect url is the url user will be redirected to after signing out.
+
+You can overwrite this value by passing url query param `pomerium_redirect_uri` or post value `pomerium_redirect_uri`
```
Comment (desimone) [inline docs/reference/readme.md:1023]:
```suggestion
You can overwrite this behavior by passing the query param `pomerium_redirect_uri` or post value `pomerium_redirect_uri` to the `/.pomerium/signout/` endpoint. 
```

File: go.sum (hunk 3 above)
Comment (cuonglm) [reply go.sum:273]: I think that should be a bug, or they run `go mod tidy` with some special option, causes the bot never remove old dependencies in `go.sum`.

File: docs/reference/readme.md (hunk 8 above)
Comment (cuonglm) [reply docs/reference/readme.md]: Fixed

File: config/options.go (hunk 7 above)
Comment (cuonglm) [reply config/options.go]: Fixed

**Pull request merged**